### PR TITLE
Improve interaction design of Date and time selection

### PR DESCRIPTION
**Pull request by dorianjp**
Trello:
https://trello.com/c/bNfAPtLt/704-improve-interaction-design-of-date-and-time-selection

Updated form wizard for creating outbreak notifications
* refactored FormView into standard View to simplify complex form logic and context processing
* Enabled multiple submit modes depending on form button submitted
* Enabled time range selection for same-day notification window timeframes
* Save notification data in session with timezone data (force America/Toronto for MVP pilot)
* Validation on form submission to disallow overlapping time ranges in session and in database
* Custom form template page with bespoke logic on errors or warning view for types of ValidationError
* Updated confirm view template to show time ranges

This PR enabled new functionality for time range support whereas previously had whole day based notifications. Other impacts not handled in this PR include:
* no time ranges shown on history view for notifications
* no support for local time zone of user or outbreak location
* no validation of session data / time ranges @ time of database transaction post-> confirmation view

### CODE REVIEW

File: outbreaks/forms.py [hunk 1]
```diff
@@ -1,85 +1,109 @@
...
-        return datetime(
-            year=int(data.get(f"year_{i}", -1)),
-            month=int(data.get(f"month_{i}", -1)),
-            day=int(data.get(f"day_{i}", -1)),
-        ).replace(tzinfo=tz)
+    def get_valid_date(self, data, start_or_end="start"):
+        tz = pytz.timezone(settings.PORTAL_LOCAL_TZ)  # TODO (mvp pilot setting) Change this for multi-tz rollout
+        default_time = '0:00:00:000000' if start_or_end == "start" else '23:59:59:999999'
+        hour, minute, second, ms = [int(x) for x in data.get(f"{start_or_end}_time", default_time).split(':')]
+        return tz.localize(datetime(
```
Comment (dorianjp) [inline outbreaks/forms.py:92]: FYI ran into bugs w datetime().replace(tzinfo):
https://pythonhosted.org/pytz/

tz.localize is more consistent

File: outbreaks/templates/datetime.html [hunk 2]
```diff
@@ -1,77 +1,183 @@
...
 {% block content %}
-    <div class="content-back-link">
-        <a href="{% url 'outbreaks:search' %}" class="chevron-left">{% trans "Back" %}</a>
-    </div>
+  <div class="content-back-link">
+    <a href="{% url 'outbreaks:search' %}" class="chevron-left">{% trans "Back" %}</a>
+  </div>
+
+  {% if form.non_field_errors and show_errors %}
+    {% if warning_ind %}
```
Comment (dorianjp) [inline outbreaks/templates/datetime.html:14]: Warning is new to this project, but I had to also maintain the same style of error without using the same form logic, hence the copied and modified error ordered list below.

File: outbreaks/views.py [hunk 3]
```diff
@@ -84,130 +103,147 @@ def get_context_data(self, *args, **kwargs):
...
+        if do_post == 'add_date':  # submitting a form
+            form = DateForm(
+                post_data,
+                alert_location=request.session['alert_location'],
+                show_time_fields=post_data.get("start_time", None)
+            )
+            if form.is_valid():
+                if self.save_form_dates(form):
+                    return HttpResponseRedirect(reverse_lazy('outbreaks:datetime'))
+            show_errors = True
```
Comment (dorianjp) [inline outbreaks/views.py:164]: this is used to hide errors on the form when not really submitting it by user intent

File: outbreaks/views.py [hunk 4]
```diff
@@ -261,54 +297,45 @@ def get_context_data(self, *args, **kwargs):
...
-        num_dates = self.request.session.get("num_dates", 1)
-        context["num_dates"] = num_dates
-        context["dates"] = []
-        for i in range(num_dates):
-            dt = datetime.fromtimestamp(self.request.session[f"alert_datetime_{i}"])
-            context["dates"].append(dt.strftime(DATETIME_FORMAT))
+        context["dates"] = self.request.session['selected_dates']
         return context
 
-    @transaction.atomic
```
Comment (dorianjp) [inline outbreaks/views.py:272]: this is not necessary with the atomic block already in place below

**Comment by dorianjp** [comment]
[Improve interaction design of Date and time selection](https://trello.com/c/bNfAPtLt/704-improve-interaction-design-of-date-and-time-selection)

File: outbreaks/templates/datetime.html [hunk 5]
```diff
@@ -1,77 +1,183 @@
...
+          <td></td>
+        </tr>
+        {% for date_entry in request.session.selected_dates %}
+          <tr>
+            <td>
+              {{ date_entry.notification_txt }}
+            </td>
+            <td>
+              <a class="destructive-link"
+                 href="{% url "outbreaks:datetime_delete" forloop.counter0 %}">{% trans "Delete" %}</a>
```
Comment (dsamojlenko) [inline outbreaks/templates/datetime.html:161]: probably could use some visually-hidden text here to give context to screen readers

File: outbreaks/templates/datetime.html [hunk 6]
```diff
@@ -1,77 +1,183 @@
...
+
+    {% if request.session.selected_dates %}
+      <table>
+        <tbody>
+        <tr>
+          <th>
+            Selected dates
+          </th>
+          <td></td>
+        </tr>
```
Comment (dsamojlenko) [inline outbreaks/templates/datetime.html:153]: Should be in a `<thead>` element above `<tbody>`. Could also add a visually-hidden heading to the second column.

**Review by dsamojlenko — CHANGES_REQUESTED** [review]

File: outbreaks/templates/datetime.html (hunk 5 above)
Comment (dorianjp) [reply outbreaks/templates/datetime.html]: updated, thanks!

File: outbreaks/templates/datetime.html (hunk 6 above)
Comment (dorianjp) [reply outbreaks/templates/datetime.html]: done, thx

**Review by dsamojlenko — APPROVED** [review]
Looks good!

**Comment by EliseKa** [comment]
Finished reviewing the translations!

**Pull request merged**